2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+These cmdlets will output a warning to use the new cmdlets.
+
+### Local cache
+
+Rather than connecting to PSGallery and other repositories for finding modules,
+the user must explicitly request the current metadata of
+modules from the repositories and that is cached locally.
+`Find-PSResource` (see below) works against this local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources is ~700KB in compressed json form.
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v2` as
+that will be assumed and discovered at runtime.
+Support for local filesystem repositories must be maintained.
+A `-InstallationPolicy` switch accepts `Trusted` or `Untrusted` (default) indicating
+whether to prompt the user when installing resources from that repository.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL or installation policy.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
```

Comment:
I've also had requests to be able to install modules directly from a repository URL. Probably also makes sense as a parameter set in Save- and Find-

Reply:
I can see how that could be useful.  Will add.

Reply:
The particular ask I heard was for something like:

```powershell
Install-Module -Name MyFunModule -Repostory 'https://mypsmodulerepo.com'
```

I don't know enough about how Nuget works to weigh that up against a direct `Install-Module -FromUri $assetUri`

Reply:
The request is to support passing a repository URL, to avoid needing to register or specify trust -- that is, if I'm telling you the URL of the repository to use, trust that repository _for now_ and if necessary, register it temporarily ....

So we don't have to what we do now, which is something like this:

```posh
# I don't actually remember what github's feed urls are, bear with me:
$MyFeed = "http://feeds.Github.com/PoshCode/"
$MyFeedName = "__Temp__PoshCode__Nightly__"

if ($Repository = (Get-PSResourceRepository).Where{ $_.Url -match $MyFeed}) {
   $Repository | Set-PSResourceRepository -InstallationPolicy Trusted
} else {
   Register-PSResourceRepository -Name $MyFeedName -Url $MyFeed -InstallationPolicy Trusted
}

Install-PSResource MyModule -Resource $MyFeedName

if ($Repository) {
    $Repository | Set-PSResourceRepository # put's pack the InstallationPolicy
} else {
    Unregister-PSResourceRepository -Name $MyFeedName
}
```


Note: this is practically mandatory for installing modules from CI builds during testing ...

Reply:
Ok, this makes sense for `-Repository`.  I thought @rjmholt was asking for:

```powershell
install-module https://www.powershellgallery.com/packages/PSScriptAnalyzer/1.18.0
```